### ISSUE TITLE
Allow seting `gpu_options.allow_growth` in config

### ISSUE THREAD

**Issue by FloopCZ** [Cognexa/cxflow-tensorflow#40]
I would like to propose a discussion whether `gpu_options.allow_growth` should be set by default.

With the new multi-gpu feature, it may be very confusing that even with n_gpus set to 0 or 1, all the GPUs are immediately fully allocated even though they are never really used.

**Comment by blazekadam** [comment]
I do not think we should override TensorFlow default behavior although it may seem reasonable.

**Comment by FloopCZ** [comment]
You are right. On the other hand, simple multi-gpu support is (unfortunately) not a default tensorflow feature. I understand you point though. Any other points to discussion from @petrbel ?

**Comment by petrbel** [comment]
Could we somehow configure it from config? Maybe it can be passed to `tf.ConfigProto` in the `session_config` dict. Or we can add `model.allow_growth` config option?

I see @FloopCZ's nets hardcode this feature and it makes sense, however, overriding TF default behavior **by default** isn't ok.

So I vote for some config-related solution.

**Comment by blazekadam** [comment]
- I do not like defining config options for every TF option worth setting up, it can be achieved with `model.session_config:{allow_growth:True}` which is actually forwarded to the `ConfigProto` in `BaseModel._create_session`

- I do not see how our multi-GPU feature and TF allocation strategy is actually connected. Using TF in multi GPU environment requires careful handling of the devices (or setting up the option as you propose) regardless.

**Comment by FloopCZ** [comment]
> I do not like defining config options for every TF option worth setting up, it can be achieved with model.session_config:{allow_growth:True} which is actually forwarded to the ConfigProto in BaseModel._create_session

That would be great of course, but `allow_growth` is hidden under `gpu_options`, so I don't believe it can be set in the config. One has to override the `_create_session` to set it up.

> I do not see how our multi-GPU feature and TF allocation strategy is actually connected. Using TF in multi GPU environment requires careful handling of the devices (or setting up the option as you propose) regardless.

Very connected. Actually `n_gpus` parameter makes no sense if all the visible cards are allocated and unusable during the training. In such a case we could just hardcode `n_gpus` to the number of visible cards. Any other option will render some cards unusable without actually using them. In our tutorial we can say that you should always use `CUDA_VISIBLE_DEVICES`. Of course this will be problematic if you have some preprocessing that wants to use GPU (like OpenCV or Arrayfire or numba etc.) because it will see no GPU.

I would prefer at least some config option to set up `allow_growth`.

**Comment by blazekadam** [comment]
Implement it then. Not required in 0.2 though...

**Comment by blazekadam** [comment]
I tried to prevent tensorflow from allocating GPU memory with both:
- `device_count={'GPU':0}`
- `gpu_options.allow_growth=True`

Both options alone as well as combined resulted in allocating small amount of memory (approx. 120Mb).
Masking devices with `CUDA_VISIBLE_DEVICES` seems to be the only viable option.

Unless we want to hack around with env variables, I suggest closing this issue.

**Comment by blazekadam** [comment]
We can mask the devices for the user if `n_gpus=0` with 
```python
import os
os.putenv('CUDA_VISIBLE_DEVICES', '')
```
before creating the session (tested).

In the case of `n_gpus>0`, we can not deduce which GPU to use and some memory is allocated anyways. Perhaps we can resurrect `sch` handling the masking as a temporary workaround until we have the cxdealer.

**Comment by FloopCZ** [comment]
The trick with `putenv` sounds nice, however, it would make the free GPUs unavailable even for the dataset (which may use it through OpenCV, Eigen, ArrayFire or similar). And yes, even the `allow_growth` option always allocates approx 120MB, but the rest of the GPU is still pretty usable.

However, as you said, there is no rush to implement this. When `cxflow` gets a wider audience, this question will eventually get asked. Until then, the workaround is to override the `create_session` method as following:
```python
    def _create_session(self, session_config):
        if session_config is None:
            session_config = {}
        session_config = tf.ConfigProto(**session_config)
        session_config.gpu_options.allow_growth = True
        return tf.Session(graph=self._graph, config=session_config)
```

**Comment by FloopCZ** [comment]
@blazekadam Thank you for your research. Although I would keep this issue open and rename it to "Allow seting gpu_options.allow_growth in config".

**Comment by blazekadam** [comment]
Well, there would be already a PR for that if TF handled it properly. This 120 MB bugs me a lot though...
It's a one-liner, you are still free to implement it before we release 0.2...

**Comment by FloopCZ** [comment]
Ok, I have just tried to set:
```yaml
  session_config:
    gpu_options:
      allow_growth: true
```
And protobuf seems to do the right thing by itself. Have you encountered some issues with this @blazekadam ? We have talked about it when it was implemented and it was decided that for nested configurations such as `graph_options` and similar, the user should override `create_session_`. However, it seems that it just works. :-)
